### ISSUE TITLE
Bwd slicing with full selection; bwd slicing on opposite graph

### ISSUE THREAD

**Issue by rolyp** [explorable-viz/fluid#754]
Add the following two scenarios to our benchmarks:
- [x] Bwd slicing with all nodes selected. (This is trivial in the fwd direction, although it might be expensive to compute), because fwd of $\top$ is $\top$.) In some sense this is the closest graph analogue of `bwdEval` because it traverses all the (needed) parts of the graph, so it could be a useful baseline.
- [x] $\triangleleft_{G^{\mathsf{op}}}$ (De Morgan dual of fwd slicing as bwd slicing on opposite graph). This is important because it is the De Morgan dual of fwd slicing, and thus a component of both linked inputs and linked outputs. Performance-wise the significance is that the “fan out” of $G^{\mathsf{op}}$ may be higher than that of $G$, because it tracks all incoming dependencies (of which there may be many for some inputs), but we don’t yet have any benchmarks that exercise that (`FwdDeMorgan` computes the dual of bwd slicing, and we don’t yet benchmark any of the linking examples).
- [x] $\triangleright_{G^{\mathsf{op}}}$ (De Morgan dual of bwd slicing as fwd slicing on opposite graph).

**Comment by rolyp** [comment]
@min-nguyen And here is another potential background task for next week, also low priority.

**Comment by min-nguyen** [comment]
@rolyp 

Can you correspond $\triangleleft_{G^{\mathsf{op}}}$ (bwd slicing on opposite graph) to the implementation for me, or outline what it means to do this again? It gets a bit confusing to remember. 

I assume $(\triangleright_{G})^{\circ}$ (De Morgan dual of fwd slicing) refers to `Graph.Slice.fwdSliceDeMorgan`.


Also:
> $\triangleleft_{G^{\mathsf{op}}}$ (**bwd slicing on opposite graph**). This is important because it is the **De Morgan dual of fwd slicing** ...

I'm confused by the two bottom bullet points; $(\triangleright_{G})^{\circ}$ is different from $\triangleleft_{G^{\mathsf{op}}}$ right?

**Comment by rolyp** [comment]
It is a bit confusing. `Graph.Slice.fwdSliceDeMorgan` is fwd slicing implemented as the De Morgan dual of bwd slicing on the opposite graph, i.e. $\triangleright_{G} = (\triangleleft_{G^{\mathsf{op}}})^{\circ}$.

If we “dualise” both sides of the above equation, we get $(\triangleright_{G})^{\circ} = \triangleleft_{G^{\mathsf{op}}}$, because $-^{\circ}$ is involutive (${x^{\circ}}^{\circ} = x$). This means that we can implement the De Morgan dual of fwd slicing (needed for linked outputs/inputs) by backward slicing on $G_{\mathsf{op}}$.

What this actually means is that linking doesn’t need to use the De Morgan dual explicitly at all: it can always be implemented by backward slicing, using either $G; G^{\mathsf{op}}$ (for linked outputs) or $G^{\mathsf{op}};G$ (for linked inputs) as the underlying graph. Which makes sense because backwards slicing along $G^{\mathsf{op}}$ tells us “what depends on me”, which is all we were using the De Morgan dual of forward slicing to calculate.

**Comment by min-nguyen** [comment]
Okay, so is the following understanding correct? (I say "sources" to refer to the outputs of a program, but with the understanding that these may not be source nodes in the graph)

#### Produces ("sufficient for")
$\triangleright_{G} \equiv (\triangleleft_{G^{\mathsf{op}}})^{\circ}$: Forward slicing $\equiv$ De Morgan dual of backward slicing on the opposite graph

Given some input selections $\alpha s \subseteq \textsf{Sinks}(G)$, we implement $(\triangleleft_{G^{\mathsf{op}}})^{\circ}$ by 
   1) backward slicing $G^{\mathsf{op}}$ wrt the output selections $\textsf{Sources}(G^{\mathsf{op}})/\alpha s$ to get input selections $\beta s \subseteq \textsf{Sinks}(G^\mathsf{op})$
   2) taking the complement of resulting input selections, i.e. $\textsf{Sinks}(G^{\mathsf{op}}) / \beta s$

#### Consumes ("requires")
$\triangleleft_{G} \equiv (\triangleright_{G^{\mathsf{op}}})^{\circ}$: Backward slicing $\equiv$ De Morgan dual of forward slicing on the opposite graph

Given some output selections $\beta s \subseteq \textsf{Sources}(G)$, we implement $(\triangleright_{G^{\mathsf{op}}})^{\circ}$ by 
   1) forward slicing $G^{\mathsf{op}}$ wrt the input selections $\textsf{Sinks}(G^{\mathsf{op}})/\beta s$ to get output selections $\alpha s  \subseteq \textsf{Sources}(G^\mathsf{op})$
   2) taking the complement of resulting output selections, i.e. $\textsf{Sources}(G^{\mathsf{op}}) / \alpha s$,


#### Consumed by ("what depends on me")
$(\triangleright_{G})^{\circ} \equiv \triangleleft_{G^{\mathsf{op}}}$: De Morgan dual of forward slicing $\equiv$ Backward slicing on the opposite graph

Given some input selections $\alpha s \subseteq \textsf{Sinks}(G)$, we implement $\triangleleft_{G^{\mathsf{op}}}$ by backward slicing ${G^{\mathsf{op}}}$ wrt the output selections $\alpha s \subseteq \textsf{Sources}(G^{\mathsf{op}})$ 

#### Solely consumes ("depends uniquely on")
$(\triangleleft_{G})^{\circ} \equiv \triangleright_{G^{\mathsf{op}}}$: De Morgan dual of backward slicing $\equiv$ Forward slicing on the opposite graph

Given some output selections $\beta s \subseteq \textsf{Sources}(G)$, we implement $\triangleright_{G^{\mathsf{op}}}$ by forward slicing ${G^{\mathsf{op}}}$ wrt the input selections $\beta s \subseteq \textsf{Sinks}(G^{\mathsf{op}})$

**Comment by rolyp** [comment]
Spot on, except that you need to swap the names of the first two relations:
- “consumes” (forward slicing) should be “produces” (what we’ve also called “sufficient for”)
- “produces” should be “consumes”

This makes intuitive sense because if we keep the direction of the analysis fixed (e.g. $\triangleleft$) and just swap $G$ for $G^{\mathsf{op}}$, we just get the inverse relation (“consumed by” rather than “consumes”).

I think the case for #760 is quite strong as it should allow us to represent some of these relationships directly.

**Comment by min-nguyen** [comment]
I've implemented the duals as simply:

```haskell
bwdSliceDual :: forall g. Graph g => Set Vertex -> g -> g
bwdSliceDual αs0 g0 = fwdSlice αs0 (op g0)

fwdSliceDual :: forall g. Graph g => Set Vertex -> g -> g
fwdSliceDual αs0 g0 = bwdSlice αs0 (op g0)
```

Was this what you were imagining, e.g. are there any alternative versions to be implemented? From reading the original comments, it seems like there's an implementation of `fwdSliceDual` which is explicitly written in terms of the De Morgan dual instead; I'm not sure what this would look like. Likewise with the explicit De Morgan implementation of `bwdSliceDual`.

**Comment by rolyp** [comment]
@min-nguyen The other implementation of `fwdSliceDual` would follow the pattern of `fwdSliceDeMorgan`, assuming the latter were fixed to apply the final negation. As it stands, `fwdSliceDeMorgan` negates the input but not the output:

```purescript
-- Doesn't do the final negation..
fwdSliceDeMorgan :: forall g. Graph g => Set Vertex -> g -> g
fwdSliceDeMorgan αs_0 g_0 =
   bwdSlice (sinks g_0 `difference` αs_0) (op g_0)
```

The final negation is handled  in `testGraph` (by `<#> not`). Cleaning this up so that the dual is implemented in a single place, using something like the `deMorgan` helper I posted the other day:

```purescript
deMorgan :: forall a b. BooleanLattice a => BooleanLattice b => Endo (a -> b)
deMorgan = (neg >>> _) >>> (_ >>> neg)
```

is one of the goals of #764. A subtask will be to be explicit about the powerset implementation of `BooleanLattice` (which will require keeping the original ambient set around in order to calculate the complement).

Once this is done, we can express `fwdSliceDual` as `deMorgan bwd`, as per `GaloisConnection.dual`.

**Comment by rolyp** [comment]
@min-nguyen Ok, so I’ve pushed a bit more on the new abstractions that should help us make the structure of the problem domain more explicit in the code. I’ve introduced `BoolAlg` as a candidate replacement for the `BooleanLattice` hierarchy (“Boolean algebra” and “Boolean lattice” are synonyms btw) and have changed the De Morgan helpers in `GaloisConnection` (currently unused) to go via `BoolAlg`. The design pattern I’m exploring is to use records rather than type classes (similar to what you might do in Agda or Lean, but without the ability to use the records at the type level) because that gives us the flexibility we need, e.g. to define a `GaloisConnection` per $e, \gamma$ pair, or a `BoolAlg` for the slices of a specific $v$.

My impression so far is that this will improve things quite a bit, although it’s early days so take a look and see what you think. If you’re interested in working on this you could start thinking about how to represent desugaring as a `GaloisConnection`.

**Comment by rolyp** [comment]
Initial pass over this done. We'll come back to more performance evaluation once we've completed #780 and #779.
